clouddriver-security/src/main/java/com/netflix/spinnaker/clouddriver/security/BaseProvider.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2020 Armory
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.clouddriver.security;
+
+import com.netflix.spinnaker.cats.agent.Agent;
+import com.netflix.spinnaker.cats.agent.AgentScheduler;
+import com.netflix.spinnaker.cats.agent.AgentSchedulerAware;
+import com.netflix.spinnaker.cats.module.CatsModule;
+import com.netflix.spinnaker.cats.module.CatsModuleAware;
+import com.netflix.spinnaker.cats.provider.Provider;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public abstract class BaseProvider extends AgentSchedulerAware implements Provider {
```

Comment:
I agree on this very case (a provider) but I still want to add a side note:

We don't need to support extending explicitly but we also don't need to prevent others from modifying either by using `final` everywhere. If we were writing a library instead of product like Spinnaker, I'd say why not! In reality, we often have to fix older releases for users who cannot always upgrade to the latest for a variety of reasons. More often than not, back-porting a fix is not feasible. So we either fork (boo!) or we fix by injecting changes and it's harder to do when everything is final.

That's why generally speaking, I prefer to leave things as open as possible and assume developers read the code they're extending. `private` shows intent, `final` is hostile! There are exceptions of course such as on a performance sensitive code path.

Reply:
I agree in general.  The reason I've been trying to be more explicit about what should be extended and what shouldn't is because `clouddriver` has historically been between an end-user application and a library.  On the one hand we primarily distribute and support the end-user application, but on the other hand we have often [encouraged people](https://spinnaker.io/guides/developer/extending/crd-extensions/) to write extensions on top of `clouddriver` as a library.

That has at times made it difficult for maintainers as it's not always clear when changing a class whether internal contracts are actually external because people have extended them, which is my motivation for trying to add `final` to places where we really don't expect extensions.  That being said probably plugins will fully solve this problem, as then we'll be able to say that anything not explicitly indicated as an extension point has no guarantee of backwards-compatibility.